### PR TITLE
chore(NODE-4018): add doc generation scripts and template to main

### PR DESCRIPTION
**Pull request by baileympearson**
#### What is changing?

This PR adds scripts needed to generate API docs into the main branch.

- The Hugo static site template used to host the docs is included in this PR.  The only file that was changed in the site template is `etc/docs/template/data/releases.toml`.  Everything else in `etc/docs/template` can be ignored.
- Two new packages were installed as dev dependencies.  I installed typedoc (because we already used it in our build:docs script, but it wasn't a dependency) and I installed a toml parser / serializer, used to update the static site template data when generating docs.
- The contents of `docs` was moved.
  - The images and charts were moved into the `etc` folder.
  - The `.md` files were moved into the top level of the repo.

Once this PR merges, a second PR will copy the contents of `gh-pages` to the `docs` folder.

Link to sample docs: https://baileympearson.github.io/node-mongodb-native/.

##### Is there new documentation needed for these changes?

No.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: etc/docs/template/data/releases.toml
Comment (dariakp) [inline etc/docs/template/data/releases.toml:9]: I noticed that this is `semvar` instead of `semver` is that how it was originally? what were the changes that were made to this file vs the original one?

File: package.json
Comment (dariakp) [inline package.json:81]: I recall Neal mentioning that this didn't play well with npm due to strict TS version requirements, is this no longer the case?

File: etc/docs/template/README.md
Comment (dariakp) [inline etc/docs/template/README.md:1]: why is there stuff about the java driver here?

File: etc/docs/README.md
Comment (dariakp) [inline etc/docs/README.md:19]: this was added to dev dependencies, will it need to be installed separately, too, or should it not be in dev dependencies?

File: etc/docs/README.md
Comment (dariakp) [inline etc/docs/README.md:32]:
```suggestion
#### Generate Documentation for the New Version
```

File: etc/docs/generate-docs.ts
Comment (dariakp) [inline etc/docs/generate-docs.ts:51]: should this be semver?

File: etc/docs/README.md
Comment (dariakp) [inline etc/docs/README.md:52]:
```suggestion
1. Generate the new API docs using typedoc
```

File: etc/docs/README.md
Comment (dariakp) [inline etc/docs/README.md:45]: Should this just say "commit the changes", create a PR, and once it's merged into main, that will update the docs?

File: etc/docs/README.md
Comment (dariakp) [inline etc/docs/README.md:54]: I don't think we should skip the PR step here

File: etc/docs/generate-docs.ts
Comment (dariakp) [inline etc/docs/generate-docs.ts:38]: Why are we switching this from `<version> Driver` to `Driver <version>`? Won't it look a bit jarring on the index page? Also, if we want to default to generating for the "next" driver version `Next Driver` would probably read better than `Driver Next`, so if we end up taking an argument for this ``` `${semVer} Driver` ``` would probably be better to use as the standard than the other way around

File: etc/docs/README.md
Comment (dariakp) [inline etc/docs/README.md:39]: can we make this not be a manual step but instead a command line arg to an npm script (preferably in package.json, too)? that would make it easier to eventually generate docs for automatic checking in the CI

File: etc/docs/README.md
Comment (dariakp) [inline etc/docs/README.md:53]: the script already does the copying into docs

File: etc/docs/generate-docs.ts
Comment (dariakp) [inline etc/docs/generate-docs.ts:41]: what does this affect? I didn't see a difference in the TOC based on this

File: etc/docs/generate-docs.ts
Comment (dariakp) [inline etc/docs/generate-docs.ts:48]: I noticed this always links to latest docs instead of that version's specific docs, is that intentional? might be a question for @nbbeeken

File: etc/docs/generate-docs.ts
Comment (dariakp) [inline etc/docs/generate-docs.ts:87]: Should this allow for updating the docs that were already generated for a given version? I think at least for main/"next", it would be nice to regenerate and update docs continuously instead of just the one time

**Review by dariakp — CHANGES_REQUESTED** [review]

File: etc/docs/template/data/releases.toml
Comment (baileympearson) [reply etc/docs/template/data/releases.toml:9]: no that's just a typo.  I added semverVersion to each line so that each entry matched the schema in my script but that's the only change in this file

File: package.json
Comment (baileympearson) [reply package.json:81]: I thought that was jsdoc - used to generate the legacy documentation.  @nbbeeken do you remember which package caused problems?

File: etc/docs/generate-docs.ts
Comment (baileympearson) [reply etc/docs/generate-docs.ts:41]: It's used in the html template - https://github.com/mongodb-js/learn-mongodb-docs/blob/master/site/layouts/partials/releases.html

File: etc/docs/generate-docs.ts
Comment (baileympearson) [reply etc/docs/generate-docs.ts:48]: I'm not sure.  My schema in the script simply matches the existing toml data used to generate our static site.  If looking into this is something we want to include in this work, I can investigate

File: package.json
Comment (nbbeeken) [reply package.json:81]: It was [typedoc](https://github.com/TypeStrong/typedoc/blob/master/package.json#L30-L32). When the next version of typescript drops we get install errors until typedoc adopts it, which we usually don't need to be blocked on

File: etc/docs/generate-docs.ts
Comment (baileympearson) [reply etc/docs/generate-docs.ts:87]: It could - I'm happy to update the script to support that if we wanted.  FYI though, although the script supports "main/next" simply because you can name your version whatever you like, it would take a bit more effort to make it work well.  If that's something we want to support in this PR, I can also make those changes

File: etc/docs/generate-docs.ts
Comment (baileympearson) [reply etc/docs/generate-docs.ts:38]: This was unintentional.  I'll flip the wording

File: etc/docs/generate-docs.ts
Comment (dariakp) [reply etc/docs/generate-docs.ts:41]: oh, it looks like it gets turned into a css class, probably worth mentioning or enumerating what classes are available/meaningful; I feel like maybe @nbbeeken was using it to control whether the older docs are displayed or collapsed

File: etc/docs/generate-docs.ts
Comment (dariakp) [reply etc/docs/generate-docs.ts:48]: I was hoping Neal had a quick answer there

File: etc/docs/generate-docs.ts
Comment (dariakp) [reply etc/docs/generate-docs.ts:87]: Just curious - what would be the issues we'd need to work around with using something like `Next` instead of a real version? Mainly I'm just concerned with always being able to properly preview the current docs and to generate them in a consistent spot so that they *can* be included in PRs

File: etc/docs/template/README.md
Comment (baileympearson) [reply etc/docs/template/README.md]: I'm not sure.  I removed it because the readme didn't seem necessary anymore

File: etc/docs/generate-docs.ts
Comment (baileympearson) [reply etc/docs/generate-docs.ts]: I assume we'd always want `main` at the top of the list of driver versions.  So basically just add logic to detect when `main` is passed as the version, and make sure to place it at the top of the list driver versions on the generated static site

File: etc/docs/template/README.md
Comment (nbbeeken) [reply etc/docs/template/README.md]: Isn't Java a strictly typed version of JavaScript? 🤪  I think our docs was forked from Java docs long ago, our [404](https://mongodb.github.io/node-mongodb-native/ada) page still has Duke on it

File: etc/docs/template/static/s/css/frontpage.css
Comment (nbbeeken) [inline etc/docs/template/static/s/css/frontpage.css:235]: Let's change this to: https://wiki.corp.mongodb.com/pages/viewpage.action?pageId=79305748&preview=/80089640/133270877/node.png :)

File: etc/docs/template/layouts/partials/assets/analytics.html
Comment (nbbeeken) [inline etc/docs/template/layouts/partials/assets/analytics.html:7]: Can we find out who has access to this reporting, we should double check this goes somewhere useful.

File: etc/docs/template/layouts/partials/assets/javascripts.html
Comment (nbbeeken) [inline etc/docs/template/layouts/partials/assets/javascripts.html:3]: Do these get created by the generation of the site? I'm not seeing any of the `/lib/` scripts in this PR

File: etc/docs/template/layouts/partials/header/main.html
Comment (nbbeeken) [inline etc/docs/template/layouts/partials/header/main.html:6]: We should probably change this:
https://www.mongodb.com/brand-resources

File: etc/docs/template/layouts/partials/footer.html
Comment (nbbeeken) [inline etc/docs/template/layouts/partials/footer.html:3]: I think we still hold the copyright :) so this could be 2022 (or maybe we don't need this?)

File: etc/docs/template/layouts/404.html
Comment (nbbeeken) [inline etc/docs/template/layouts/404.html:22]: Just a quick win idea, a link back to the homepage or mongodb manual could be good here, if we end up having to break legacy version links down the road this page may get an uptick in visits.

File: etc/docs/template/static/s/js/jquery.js
Comment (nbbeeken) [inline etc/docs/template/static/s/js/jquery.js:2]: Should we be concerned about this? If we could just try the latest jquery still works, it might be a good idea to update

File: etc/docs/template/layouts/index.html
Comment (nbbeeken) [inline etc/docs/template/layouts/index.html:5]: The icon should be updated: https://www.mongodb.com/assets/images/global/favicon.ico

File: etc/docs/template/.gitignore
Comment (nbbeeken) [inline etc/docs/template/.gitignore:1]: Is this relevant to this directory?

File: etc/docs/generate-docs.ts
Comment (nbbeeken) [inline etc/docs/generate-docs.ts:83]: `await writeFile` instead?

File: etc/docs/legacy-generate.sh
Comment (nbbeeken) [inline etc/docs/legacy-generate.sh:101]: We decided to do this in a follow up right? should we ticket the TODO?

File: package.json
Comment (nbbeeken) [inline package.json:124]: Totally fine if you think this is not the right direction but: https://github.com/nbbeeken/nbbeeken.github.io/blob/main/scripts/server.js we could have a little server that works just on node to avoid the need for python here. I guess we all need python for mlaunch so not a huge deal.

File: typedoc.json
Comment (nbbeeken) [inline typedoc.json:6]: We want to exculdeInternal, so this should be true right?

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: etc/docs/template/layouts/partials/assets/javascripts.html
Comment (baileympearson) [reply etc/docs/template/layouts/partials/assets/javascripts.html:3]: No.  These scripts are present in the gh-pages branch (https://github.com/mongodb/node-mongodb-native/tree/gh-pages/s/lib).  They'll be present after part 2 of this PR, when I move the contents of `gh-pages` to main.

File: etc/docs/legacy-generate.sh
Comment (baileympearson) [reply etc/docs/legacy-generate.sh:101]: What exactly are you referring to here?  

I included this script only so that, if needed, we could generate docs for legacy versions (3.x).  The other script I wrote should support generating docs for any 4x version

File: etc/docs/template/static/s/css/frontpage.css
Comment (baileympearson) [reply etc/docs/template/static/s/css/frontpage.css:235]: Let's update this in my second PR - the `img` files don't exist in this PR since there are only changes to the site template (which in our case, doesn't include the images).

File: etc/docs/template/layouts/partials/assets/analytics.html
Comment (baileympearson) [reply etc/docs/template/layouts/partials/assets/analytics.html:7]: Yeah, I can check

File: etc/docs/README.md
Comment (baileympearson) [inline etc/docs/README.md:35]: This step hasn't been fully tested because the actual documentation site won't be ported over until a follow up PR.  I'll test this and ensure that the preview step works as expected then.

File: etc/docs/generate-docs.ts
Comment (baileympearson) [reply etc/docs/generate-docs.ts]: I can't find any css classes that correspond to the statuses we use now and I confirmed on the current hosted version of our docs that these css classes don't actually apply any styles.  I'll remove the code in the template that applies the styles.  

This value is also used to filter out legacy versions of the driver when building the template though.

File: etc/docs/template/layouts/partials/assets/analytics.html
Comment (baileympearson) [reply etc/docs/template/layouts/partials/assets/analytics.html:7]: Mary confirmed that we do intentionally collect these analytics

File: etc/docs/template/layouts/partials/header/main.html
Comment (baileympearson) [reply etc/docs/template/layouts/partials/header/main.html:6]: I'll address this with the other `img` related comments in the follow up PR!

File: etc/docs/template/layouts/index.html
Comment (baileympearson) [reply etc/docs/template/layouts/index.html:5]: same comment as other `img` changes

File: etc/docs/template/layouts/404.html
Comment (baileympearson) [reply etc/docs/template/layouts/404.html:22]: Do you mind if I make this change in my follow up PR?  There's no way to test that the link I add works with the changes in this PR (since we don't actually have the whole site included).

File: etc/docs/template/static/s/js/jquery.js
Comment (baileympearson) [reply etc/docs/template/static/s/js/jquery.js:2]: Do you mind if I make this change in my follow up PR? There's no way to test that the link I add works with the changes in this PR (since we don't actually have the whole site included).

File: etc/docs/README.md
Comment (dariakp) [inline etc/docs/README.md:37]: Just a question - is there an actual github action that *we* configure to rebuild the site or does it just automatically get updated once we point the github pages setting in the repo to the main docs branch? If there is no action for us to manually manage, it may be better just to say here that our github pages config points to the /docs directory on main and is kept up to date by github itself

File: typedoc.json
Comment (dariakp) [inline typedoc.json:6]: with this being the right setting, I don't think we should keep the line above this one to `"excludeTags": ["internal"]`, since presumably they wouldn't be present in the first place, and if we somehow accidentally lose this flag, excluding the internal tag from being displayed would just make it more confusing to our users

File: etc/docs/generate-docs.ts
Comment (dariakp) [reply etc/docs/generate-docs.ts]: In that case, should we make it an enum of `'current' | 'not-supported'` and/or add validation that it's not an unknown value (to prevent typos)?

File: etc/docs/generate-docs.ts
Comment (dariakp) [inline etc/docs/generate-docs.ts:81]: I think we could make this process a bit more resilient to generating versions out of order by just writing a custom semver sort function and passing it to both tomlData and and jsonVersions (should be straightforward since it's internal and also we only do major.minor, no patch or pre-release flags); we could also have it check for our special keyword `main` or `current`, something like:
```js
const versionSort = (a, b) => {
  const vA = a.version;
  const vB = b.version;
  if (vA === 'current') { return -1; }
  if (vB === 'current') { return 1; }
//  < logic to return -1 if vA > vB and 1 if vA < bB >
// since we want the most recent version first
}

<unshift new version>
tomlData.versions.sort(versionSort);
tomlData.current = tomlData.versions[0].versions;

<unshift new version>
jsonVersions.sort(versionSort);

```

**Review by dariakp — CHANGES_REQUESTED** [review]

File: etc/docs/generate-docs.ts
Comment (dariakp) [reply etc/docs/generate-docs.ts]: I have a suggestion in a new comment to handle this with a sort, let me know what you think

File: package.json
Comment (nbbeeken) [reply package.json:124]: this is fine to leave as is, can we fix the spacing on the args? `python  -m` -> `python -m`

File: etc/docs/generate-docs.ts
Comment (nbbeeken) [reply etc/docs/generate-docs.ts]: We don't currently use the supported class but my quick fix to make the version list table get too long isn't ideal, I think making use of these CSS classes to better indicate the status of each version is a future improvement. Is it okay if we keep this and consider making use of it in the next PR or just future work?

I'm thinking background color like [wikipedia uses](https://en.wikipedia.org/wiki/OpenSSL#Major_version_releases)

File: package.json
Comment (baileympearson) [reply package.json:124]: Why not two spaces everywhere instead of one?  `python  -m  http.server  --directory  etc/docs/build`?  🤔 

jk, good catch! I'll remove the space

File: etc/docs/README.md
Comment (baileympearson) [reply etc/docs/README.md:37]: The Github action is managed by Github and it should update as soon as we switch the location of the static site hosting from `gh-pages` to `main/docs`.  I'll update the wording here

File: etc/docs/generate-docs.ts
Comment (nbbeeken) [reply etc/docs/generate-docs.ts]: It's always linked to the latest, we could add versioning but not here I think.

**Review by dariakp — DISMISSED** [review]
Approved with the expectation of a follow up with the full docs and another with the template/generating script updates

**Review by nbbeeken — APPROVED** [review]

**Pull request merged**
